Modules/Quotas/src/Quotas.Application/Identities/Queries/GetIdentity/GetIdentityResponse.cs

```diff
@@ -1,36 +1,53 @@
 ﻿using Backbone.Modules.Quotas.Application.DTOs;
 using Backbone.Modules.Quotas.Domain.Aggregates.Identities;
 using Backbone.Modules.Quotas.Domain.Aggregates.Metrics;
+using Backbone.Modules.Quotas.Domain.Metrics;
 
 namespace Backbone.Modules.Quotas.Application.Identities.Queries.GetIdentity;
 public class GetIdentityResponse
 {
-    public GetIdentityResponse(string identityAddress, IEnumerable<TierQuota> tierQuotas, IEnumerable<IndividualQuota> individualQuotas, IEnumerable<Metric> metrics)
+    public string Address { get; set; }
+    public IEnumerable<QuotaDTO> Quotas { get; set; }
+}
+
+public class GetIdentityResponseBuilder
+{
+    private readonly MetricCalculatorFactory _metricCalculatorFactory;
+
+    public GetIdentityResponseBuilder(MetricCalculatorFactory metricCalculatorFactory)
+    {
+        _metricCalculatorFactory = metricCalculatorFactory;
+    }
+
+    public async Task<GetIdentityResponse> BuildResponse(string identityAddress, IEnumerable<TierQuota> tierQuotas, IEnumerable<IndividualQuota> individualQuotas, IEnumerable<Metric> metrics, CancellationToken cancellationToken)
     {
         var quotasList = new List<QuotaDTO>();
-        quotasList.AddRange(individualQuotas.Select(q =>
-            new QuotaDTO(
-                q.Id,
-                QuotaSource.Individual,
-                new MetricDTO(metrics.First(m => m.Key == q.MetricKey)),
-                q.Max,
-                q.Period.ToString()
-            )
-        ));
-        quotasList.AddRange(tierQuotas.Select(q =>
-            new QuotaDTO(
-                q.Id,
-                QuotaSource.Tier,
-                new MetricDTO(metrics.First(m => m.Key == q.MetricKey)),
-                q.Max,
-                q.Period.ToString()
-            )
-        ));
-
-        Address = identityAddress;
-        Quotas = quotasList;
+
+        var allQuotas = (individualQuotas as IEnumerable<Quota>).Union(tierQuotas);
+
+        foreach (var quota in allQuotas)
+        {
+            var usage = await CalculateUsageAsync(_metricCalculatorFactory, quota, identityAddress, cancellationToken);
+            quotasList.Add(new QuotaDTO(
+                    quota.Id,
+                    quota is TierQuota ? QuotaSource.Tier : QuotaSource.Individual,
+                    new MetricDTO(metrics.First(m => m.Key == quota.MetricKey)),
+                    usage,
+                    quota.Max,
+                    quota.Period.ToString()
+                ));
```

Comment:
How about

```suggestion
            quotasList.Add(new QuotaDTO(
                    quota,
                    metrics.First(m => m.Key == quota.MetricKey),
                    usage,
                ));
```

This
1. makes the mapping of `Source` mapping reusable
2. simplifies the usage

Reply:
This would also pollute the DTO with unwanted attributes. Perhaps we could use a `Create` method like before?

Reply:
You don't map all the properties of the Quota. You can just choose the ones you like. Sure, you have access to all of them, but that's no problem in my opinion.